realm/rfc/0001-backlink-query-syntax.md

```diff
@@ -0,0 +1,232 @@
+  can point to it. This is not supported by Core, but it is a natural extension.
+
+- Should support backlinks from multiple fields.
+
+- Should be usable from the Dynamic API.
+
+- Should either support polymorphic objects or not prevent them from being 
+  implemented.
+
+- Ideally it should work for both a pure text-based variant as well as the 
+  current mixed style found in Realm Java.
+  
+
+# Detailed design
+
+**Example model classes:**
+
+```
+public class Person extends RealmObject {
+	private Dog favoriteDog;
+	private RealmList<Dog> dogs;
+}
+
+public class Dog extends RealmObject {
+	private String name
+
+	// Named backlinks below
+	@Backlink({"favoriteDog"})
+	private RealmResults<Person> favoriteOwner;
+
+	@Backlink({"favoriteDog", "dogs"})
+	private RealmResults<Person> allOwners;
+}
+```
+
+**Introduce the following query extension**
+
+- Use `"y[x.z]"` to search field Y on type X where X is a backlink through 
+  field Z.
+- Searching multiple fields should be done in individual query methods:
+  `equalTo("y[x.z]", "foo").equalTo("y[x.z1]", "foo")`
+- This is a supplement to defining the named backlinks.
+- No changes to the migration API.
+- No changes to the schema definition.
+
+Example:
+
+```
+// Named backlink queries
+realm.where(Dog.class).equalTo("allOwners.name", "John").findAll();
+realm.where(Dog.class).equalTo("favoriteOwner.name", "John").findAll();
+
+// Unnamed backlink queries
+realm.where(Dog.class).equalTo("name[Person.favoriteDog]", "John").findAll();
+realm.where(Dog.class)
+  .equalTo("name[Person.favoriteDog]", "John")
+  .equalTo("name[Person.dogs]", "John")
+  .findAll();
+
+// Example in a purely text based variant
+"name[Person.favoriteDog] = 'John'"
+```
+
+**Implementation**
+Implementation is straight forward as we have all schema information available 
+when parsing queries. So simple String manipulation will make it possible to 
+split a field name into a name and a type part.
+
+Validating that the type is an actual backlink with the given field does require 
+an `O(n)` lookup where `m: number of fields in Person`. Since number of fields 
+are generally low, this shouldn't be a problem in practise.
+
+**Advantages**
+- Using "[]" is not an allowed field characters in any of our supported 
+  languages as far as I know. So parsing the above expressions should be 
+  possible with no ambiguity.
+
+- The default case is easy to describe.
+
+- Using [] as a modifier already has precedent in NSPredicate.
+
+
+**Disadvantages**
+
+- Unknown syntax. This concept is not used nor supported in neither NSPredicate, 
+  LINQ or SQL. It will be a "power feature", and might make the language feel 
+  complicated and hard to use.
+
+- Moves the Java query language further away from being type safe.
+
+- Will introduce two ways of querying backlinks.
+
+- Without wildcards for unnamed backlinks queries risk getting extremely verbose.
+
```

Comment:
Disadvantage: no way to understand what the syntax means for a new user.
Disadvantage: syntax using only special characters is very hard to search for documentation on.